utils/fs/memory/memory.go

```diff
@@ -0,0 +1,333 @@
+
+func (fs *Memory) TempFile(dir, prefix string) (fs.File, error) {
+	filename := fmt.Sprintf("%s_%d", prefix, time.Now().UnixNano())
+	fullpath := fs.Join(fs.base, dir, filename)
+	return fs.Create(fullpath)
+}
+
+func (fs *Memory) Rename(from, to string) error {
+	from = fs.Join(fs.base, from)
+	to = fs.Join(fs.base, to)
+
+	if _, ok := fs.s.files[from]; !ok {
+		return os.ErrNotExist
+	}
+
+	fs.s.files[to] = fs.s.files[from]
+	fs.s.files[to].BaseFilename = to
+	delete(fs.s.files, from)
+
+	return nil
+}
+
+func (fs *Memory) Remove(filename string) error {
+	if _, err := fs.Stat(filename); err != nil {
+		return err
+	}
+
+	fullpath := fs.Join(fs.base, filename)
+	delete(fs.s.files, fullpath)
+	return nil
+}
+
+func (fs *Memory) clean(path string) string {
+	if len(path) <= 1 {
+		if path != string(separator) {
+			return path
+		}
+
+		return ""
+	}
+
+	if path[0] == separator {
+		path = path[1:]
+	}
+
+	l := len(path)
+	if path[l-1] == separator {
+		path = path[:l-1]
+	}
+
+	return path
+}
+
+func (fs *Memory) Join(elem ...string) string {
+	return filepath.Join(elem...)
+}
+
+func (fs *Memory) Dir(path string) fs.Filesystem {
+	return &Memory{
+		base: fs.Join(fs.base, path),
+		s:    fs.s,
+	}
+}
+
+func (fs *Memory) Base() string {
+	return fs.base
+}
+
+type file struct {
+	fs.BaseFile
+
+	c     *content
+	p     int64
+	flags int
+}
+
+func newFile(base, fullpath string, flags int) *file {
+	filename, _ := filepath.Rel(base, fullpath)
+
+	return &file{
+		BaseFile: fs.BaseFile{BaseFilename: filename},
+		c:        &content{},
+		flags:    flags,
+	}
+}
+
+func (f *file) Read(b []byte) (int, error) {
+	if f.IsClosed() {
+		return 0, fs.ErrClosed
+	}
+
+	if f.flags&os.O_RDWR != 0 && f.flags&os.O_RDONLY != 0 {
+		return 0, errors.New("read not supported")
+	}
+
+	n, err := f.c.ReadAt(b, f.p)
+	f.p += int64(n)
+
+	return n, err
+}
+
+func (f *file) Seek(offset int64, whence int) (int64, error) {
+	switch whence {
+	case io.SeekCurrent:
+		if offset == 0 {
+			return f.p, nil
+		}
+
+		f.p += offset
+	case io.SeekStart:
+		f.p = offset
+	case io.SeekEnd:
+		f.p = f.c.size - offset
+	}
+
+	return f.p, nil
+}
+
+func (f *file) Write(p []byte) (int, error) {
+	if f.IsClosed() {
+		return 0, fs.ErrClosed
+	}
+
+	if f.flags&os.O_RDWR != 0 && f.flags&os.O_WRONLY != 0 {
+		return 0, errors.New("read not supported")
+	}
+
+	n, err := f.c.WriteAt(p, f.p)
+	f.p += int64(n)
+
+	return n, err
+}
+
+func (f *file) Close() error {
+	f.Closed = true
+	return nil
+}
+
+func (f *file) Open() error {
+	f.Closed = false
+	return nil
+}
+
+type fileInfo struct {
+	name  string
+	size  int64
+	isDir bool
+}
+
+func newFileInfo(base, fullpath string, size int64) *fileInfo {
+	filename, _ := filepath.Rel(base, fullpath)
+
+	return &fileInfo{
+		name: filename,
+		size: size,
+	}
+}
+
+func (fi *fileInfo) Name() string {
+	return fi.name
+}
+
+func (fi *fileInfo) Size() int64 {
+	return fi.size
+}
+
+func (fi *fileInfo) Mode() os.FileMode {
+	return os.FileMode(0)
```

Comment:
I though that storing and retuning modes, may run on confusions because doesn't have any effect